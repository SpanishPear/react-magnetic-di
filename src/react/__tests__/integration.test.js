/* eslint-env jest */

import React, { Component, Fragment } from 'react';
import { mount } from 'enzyme';
import { di, DiProvider, mock } from '../../index';

const Wrapper = ({ children }) => children;
const Text = () => 'original';

class Label extends Component {
  render() {
    const [_Wrapper, _Text] = di([Wrapper, Text]);
    return (
      <_Wrapper>
        <_Text />
      </_Wrapper>
    );
  }
}

class Input extends Component {
  render() {
    const [_Text] = di([Text]);
    return <_Text />;
  }
}

const TextMock = mock(Text, () => 'mock');
const WrapperMock = mock(Wrapper, ({ children }) => children);

describe('Integration', () => {
  it('should return real dependencies if provider less', () => {
    const wrapper = mount(
      <Fragment>
        <Label />
        <Input />
      </Fragment>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should override all dependencies of same type', () => {
    const wrapper = mount(
      <DiProvider use={[TextMock]}>
        <Label />
        <Input />
      </DiProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should allow override composition', () => {
    const wrapper = mount(
      <DiProvider use={[WrapperMock]}>
        <DiProvider use={[TextMock]}>
          <Label />
          <Input />
        </DiProvider>
      </DiProvider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });
});
