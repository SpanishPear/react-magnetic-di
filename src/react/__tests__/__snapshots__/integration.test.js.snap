// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration should allow override composition 1`] = `
"<DiProvider use={{...}}>
  <DiProvider use={{...}}>
    <Label>
      <diMock(Wrapper)>
        <diMock(Text)>
          mock
        </diMock(Text)>
      </diMock(Wrapper)>
    </Label>
    <Input>
      <diMock(Text)>
        mock
      </diMock(Text)>
    </Input>
  </DiProvider>
</DiProvider>"
`;

exports[`Integration should override all dependencies of same type 1`] = `
"<DiProvider use={{...}}>
  <Label>
    <Wrapper>
      <diMock(Text)>
        mock
      </diMock(Text)>
    </Wrapper>
  </Label>
  <Input>
    <diMock(Text)>
      mock
    </diMock(Text)>
  </Input>
</DiProvider>"
`;

exports[`Integration should return real dependencies if provider less 1`] = `
"<Label>
  <Wrapper>
    <Text>
      original
    </Text>
  </Wrapper>
</Label>


<Input>
  <Text>
    original
  </Text>
</Input>"
`;
